### DISCUSSION TITLE
Proposal for defining `Species` constituents in terms of other `Species`

### DISCUSSION THREAD

**Discussion by bocklund** [pycalphad/pycalphad#300] [Ideas]
As I'm working through implementing the modified quasichemical model in the quadruplet approximation (MQMQA), I've hit a snag with pycalphad's `v.Species` objects. In the MQMQA we treat a quadruplets as constituents, like an associate, but the constituents of a quadruplet are really species and not pure elements.

For example, we can have a quadruplets of (Cu+1, Cu+1, Cl-1, Cl-1) and (Cu+2, Cu+2, Cl-1, Cl-1). If we constructed quadruplet species based on the pure element names, both quadruplets would be `v.Species('CUCUCLCL', constituents={'CU': 2.0, 'CL': 2.0})`, which would compare equal. It's important that we can access the pure element constituents to compute the mass in terms of pure elements, but it would be nice to be able to create species of species without resorting to name mangling or other hacks [1].

If we allowed a `Species().constituents` to have keys of `Species` or `str`, we could define a `v.Species().pure_constituents` property to recurse through `Species().constituents` until all the pure element leaf nodes are reached.

Another consideration is how to handle TDB files, since the `SPECIES` keywords requires defining species in terms of pure elements. We would be able to write out a valid TDB file using `Species.pure_constituents`, but a round trip `dbf.from_string(dbf.to_string())` would not preserve the relationship between species of species.

<hr>

[1] There's another issue in having two quadruplets share the same name, but have different constituents: the site fraction objects will compare equal:

```python
>>> from pycalphad import variables as v
>>> # two species share the same name, but different constituents
>>> AB = v.Species('AB', {'AB': 1})
>>> A_B = v.Species('AB', {'A': 1, 'B': 1})
>>> # check the Species equality
>>> AB == A_B
False
>>> # check the site fraction equality
>>> v.Y('PHASE', 0, AB) == v.Y('PHASE', 0, A_B)
True
```

**Comment by richardotis** [comment]
The root issue is partly theoretical in that site fractions mean something different in MQMQA than they do in the compound-energy formalism (CEF). As a practical consideration though I agree with the idea of overloading `Species` to do what you're trying to do here. I think such a design could be implemented with very minimal changes to other code. I'm particularly thinking about:

- `Model.moles()` is where all(?) mass calculation happens, so changes there should propagate cleanly
- `@property` overloading for key attributes of `Species` to abstract away any recursive definitions
- New `Species` attributes as needed to expose for MQMQA

If you're already extending TDB for the coordination number, I would also think about ways to define quadruplets that retain the round-trip property and limit ambiguity.

**Comment by bocklund** [comment]
If you use a strategy where you store the coordination numbers as parameters, changes to `Species` are not strictly necessary for the MQMQA because you can compute the mass of an element using only the quadruplet fractions and coordination numbers. You can differentiate different species within a quadruplet with some simple name mangling  (which is all hidden from an end user unless you are inspecting the symbolic expressions). I think this change would be mostly cosmetic for the MQMQA and simplify some of the name mangling. It would allow you to explore methods of storing the coordination numbers within the quadruplet species constituents as the stoichiometric ratios of that constituent. 

### API 
Here's the concept from an API usage standpoint:

```python
CL =  v.Species('CL1/+1.0')
CU1 = v.Species('CU1/+1.0')
CU2 = v.Species('CU1/+2.0')
Q_CU1_CU2_CL_CL = v.Species('CUCUCLCL', constituents={CU1: 1, CU2: 1, CL: 2})
print(Q_CU1_CU2_CL_CL.name)  # CUCUCLCL
print(Q_CU1_CU2_CL_CL.constituents)  # {Species('CU1/+1.0', 'CU1.0', charge=1): 1, Species('CU1/+2.0', 'CU1.0', charge=2): 1, Species('CL1/+1.0', 'CL1.0', charge=1): 2}
print(Q_CU1_CU2_CL_CL.pure_constituents)  # {'CU': 2.0, 'CL': 2.0}

Q_CU1_CU1_CL_CL = v.Species('CUCUCLCL', constituents={CU1: 2, CL: 2})
print(Q_CU1_CU1_CL_CL.name)  # CUCUCLCL
print(Q_CU1_CU1_CL_CL.constituents)  # {Species('CU1/+1.0', 'CU1.0', charge=1): 2, Species('CL1/+1.0', 'CL1.0', charge=1): 2}
print(Q_CU1_CU1_CL_CL.pure_constituents)  # {'CU': 2.0, 'CL': 2.0}

print(Q_CU1_CU2_CL_CL == Q_CU1_CU1_CL_CL)  # False

print(v.Y('ALPHA', 0, Q_CU1_CU2_CL_CL) == v.Y('ALPHA', 0, Q_CU1_CU1_CL_CL))  # True

C = v.Species('Q1Q2', constituents={Q_CU1_CU2_CL_CL: 2.0, Q_CU1_CU1_CL_CL: 1.0})
print(C.name)  # Q1Q2
print(C.constituents)  # {Species('CUCUCLCL', 'CL1/+1.02CU1/+1.01CU1/+2.01'): 1.0, Species('CUCUCLCL', 'CL1/+1.02CU1/+1.02'): 1.0}
print(C.pure_constituents)  # {'CU': 6.0, 'CL': 6.0}

print(C.number_of_atoms)  # 12.0
```

### Diff


And here's the diff in `Species` that gets you that behavior. Basically everywhere that you need the pure element constituents for some mass bookkeeping, you should use `Species().pure_constituents` to get the pure element strings. Obviously a PR with these changes would need to update any usages of constituents elsewhere in pycalphad, but that should be straightforward and well covered by the test suite.

```python
diff --git a/pycalphad/variables.py b/pycalphad/variables.py
index 8848816c..0596e81d 100644
--- a/pycalphad/variables.py
+++ b/pycalphad/variables.py
@@ -76,6 +76,17 @@ class Species(object):
     def __str__(self):
         return self.name

+    @property
+    def pure_constituents(self):
+        pure_constits = {}
+        for constituent, amount in self.constituents.items():
+            if isinstance(constituent, str):
+                pure_constits[constituent] = pure_constits.get(constituent, 0.0) + amount
+            elif isinstance(constituent, Species):
+                for sp_constitutent, sp_amount in constituent.pure_constituents.items():
+                    pure_constits[sp_constitutent] = pure_constits.get(sp_constitutent, 0.0) + sp_amount*amount
+        return pure_constits
+
     @property
     def escaped_name(self):
         "Name safe to embed in the variable name of complex arithmetic expressions."
@@ -84,7 +95,7 @@ class Species(object):
     @property
     def number_of_atoms(self):
         "Number of atoms per formula unit. Vacancies do not count as atoms."
-        return sum(value for key, value in self.constituents.items() if key != 'VA')
+        return sum(value for key, value in self.pure_constituents.items() if key != 'VA')

     @property
     def weight(self):
```

**Reply by richardotis** [reply]
I support this approach.
